File: engine/src/lockfile/npm.ts
import { UnsupportedLockfileError, type LockfileDep } from "./index.js";

interface NpmLockEntry {
  version?: string;
  link?: boolean;
}

interface NpmLockRoot extends NpmLockEntry {
  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
  optionalDependencies?: Record<string, string>;
  peerDependencies?: Record<string, string>;
}

/** package-lock.json v2/v3 — walk the `packages` map. v1 (npm <7) has no
 *  `packages` key and is rejected with a regenerate hint. */
export function parseNpmLockfile(
  content: string,
  manifest: Map<string, string>,
): LockfileDep[] {
  let lock: { lockfileVersion?: number; packages?: Record<string, NpmLockEntry> };
  try {
    lock = JSON.parse(content);
  } catch {
    throw new UnsupportedLockfileError("package-lock.json is not valid JSON");
  }
  if (!lock.packages || typeof lock.packages !== "object") {
    throw new UnsupportedLockfileError(
      "package-lock.json v1 is not supported — regenerate it with npm >= 7",
    );
  }

  // Direct ranges: the lockfile's root entry is authoritative; the manifest
  // fills gaps (e.g. lockfile written by an older npm).
  const root = (lock.packages[""] ?? {}) as NpmLockRoot;
  const directRanges = new Map<string, string>(manifest);
  for (const section of [
    root.dependencies,
    root.devDependencies,
    root.optionalDependencies,
    root.peerDependencies,
  ]) {
    for (const [name, range] of Object.entries(section ?? {})) {
      directRanges.set(name, range);
    }
  }

  const seen = new Map<string, LockfileDep>();
  for (const [pkgPath, entry] of Object.entries(lock.packages)) {
    if (pkgPath === "" || entry.link) continue;
    // "node_modules/foo", "node_modules/a/node_modules/@scope/b", or a
    // workspace dir ("packages/a" — skipped, it's local code, not a dep).
    const idx = pkgPath.lastIndexOf("node_modules/");
    if (idx === -1) continue;
    const name = pkgPath.slice(idx + "node_modules/".length);
    if (!name || !entry.version) continue;

    const key = `${name}@${entry.version}`;
    if (seen.has(key)) continue;
    const direct = directRanges.has(name);
    seen.set(key, {
      name,
      version: entry.version,
      direct,
      range: direct ? (directRanges.get(name) ?? null) : null,
    });
  }
  return [...seen.values()];
}

File: engine/npmguard/panel/lockfile/npm.py
"""package-lock.json v2/v3 parser.

Walks the ``packages`` map. v1 (npm < 7) has no ``packages`` key and is
rejected with a regenerate hint. A port of TS ``lockfile/npm.ts``.
"""

from __future__ import annotations

import json

from ._base import LockfileDep, UnsupportedLockfileError

_NODE_MODULES = "node_modules/"


def parse_npm_lockfile(content: str, manifest: dict[str, str]) -> list[LockfileDep]:
    try:
        lock = json.loads(content)
    except (ValueError, TypeError) as exc:
        raise UnsupportedLockfileError("package-lock.json is not valid JSON") from exc

    packages = lock.get("packages") if isinstance(lock, dict) else None
    if not isinstance(packages, dict):
        raise UnsupportedLockfileError(
            "package-lock.json v1 is not supported — regenerate it with npm >= 7"
        )

    # Direct ranges: the lockfile's root ("") entry is authoritative; the
    # manifest fills gaps (e.g. a lockfile written by an older npm).
    root = packages.get("") or {}
    direct_ranges: dict[str, str] = dict(manifest)
    if isinstance(root, dict):
        for section in (
            "dependencies",
            "devDependencies",
            "optionalDependencies",
            "peerDependencies",
        ):
            deps = root.get(section)
            if isinstance(deps, dict):
                for name, rng in deps.items():
                    if isinstance(rng, str):
                        direct_ranges[name] = rng

    seen: dict[str, LockfileDep] = {}
    for pkg_path, entry in packages.items():
        if pkg_path == "" or not isinstance(entry, dict) or entry.get("link"):
            continue
        # "node_modules/foo", "node_modules/a/node_modules/@scope/b", or a
        # workspace dir ("packages/a" — skipped, it's local code, not a dep).
        idx = pkg_path.rfind(_NODE_MODULES)
        if idx == -1:
            continue
        name = pkg_path[idx + len(_NODE_MODULES) :]
        version = entry.get("version")
        if not name or not version:
            continue

        key = f"{name}@{version}"
        if key in seen:
            continue
        direct = name in direct_ranges
        seen[key] = LockfileDep(
            name=name,
            version=version,
            direct=direct,
            range=direct_ranges.get(name) if direct else None,
        )
    return list(seen.values())
